bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/Watchdog.java

```diff
@@ -0,0 +1,240 @@
+
+    int count = 0;
+
+    while (it.hasNext()) {
+      WatchedCall<?,?> stream = it.next().getKey();
+      if (stream.cancelIfStale()) {
+        count++;
+        it.remove();
+      }
+    }
+
+    if (count > 0) {
+      LOG.warn("Found %d stale streams and cancelled them", count);
+    }
+  }
+
+  public void start(ScheduledExecutorService executor) {
+    Preconditions.checkState(scheduledFuture == null, "Already started");
+    long minTimeoutMs = Math.min(waitTimeoutMs, idleTimeoutMs);
+    long checkPeriodMs = Math.max(minTimeoutMs / 2, MIN_CHECK_PERIOD_MS);
+
+    scheduledFuture = executor.scheduleAtFixedRate(this,
+        checkPeriodMs, checkPeriodMs, TimeUnit.MILLISECONDS);
+  }
+
+  public void stop() {
+    scheduledFuture.cancel(true);
+    scheduledFuture = null;
+  }
+
+  private class WatchedCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private State state;
+
+    @GuardedBy("lock")
+    private int pendingCount = 0;
+
+    @GuardedBy("lock")
+    private long lastActivityAt = clock.currentTimeMillis();
+
+    WatchedCall(ClientCall<ReqT, RespT> delegate) {
+      super(delegate);
+    }
+
+    @Override
+    public void start(Listener<RespT> listener, Metadata metadata) {
+      ClientCall<ReqT, RespT> call = delegate();
+
+      synchronized (lock) {
+        Preconditions.checkState(state == null, "Already started");
+        state = (pendingCount == 0) ? State.IDLE : State.WAITING;
+        lastActivityAt = clock.currentTimeMillis();
+      }
+
+      openStreams.put(this, PRESENT);
+
+      call.start(new SimpleForwardingClientCallListener<RespT>(listener) {
+        @Override
+        public void onMessage(RespT message) {
+          synchronized (lock) {
+            state = State.DELIVERING;
+          }
+
+          try {
+            super.onMessage(message);
+          } finally {
+            synchronized (lock) {
+              pendingCount--;
+              lastActivityAt = clock.currentTimeMillis();
+
+              state = (pendingCount > 0) ? State.WAITING : State.IDLE;
+            }
+          }
+        }
+
+        @Override
+        public void onClose(Status status, Metadata trailers) {
+          openStreams.remove(WatchedCall.this);
+
+          // Take care to convert an error caused by StreamWaitTimeoutException into an ABORTED
+          // status. This will allow it to be retried.
+          if (status.getCause() instanceof StreamWaitTimeoutException) {
+            status = Status.ABORTED
+                .withDescription(status.getDescription())
+                .withCause(status.getCause());
+          }
+
+          super.onClose(status, trailers);
+        }
+      }, metadata);
+    }
+
+    @Override
+    public void request(int count) {
+      synchronized (lock) {
+        if (state == State.IDLE) {
+          state = State.WAITING;
+          lastActivityAt = clock.currentTimeMillis();
```

Comment:
Can `lastActivityAt` be set here even if `state != idle?`?

Reply:
No, request() should only reset the idle timeout. It should not reset the timeout for waiting on the server for a response.